File: library/src/test/java/io/github/dmitrikudrenko/logger2/LogSetupTest.java
package io.github.dmitrikudrenko.logger2;

import io.github.dmitrikudrenko.logger2.impl.AndroidHandler;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.api.mockito.PowerMockito;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;
import org.robolectric.annotation.Config;

import java.util.logging.Handler;
import java.util.logging.LogManager;
import java.util.logging.Logger;

import static org.mockito.Mockito.*;

@RunWith(PowerMockRunner.class)
@Config(manifest = Config.NONE)
@PrepareForTest({Log.class, Logger.class, LogManager.class})
public class LogSetupTest {
    private Logger logger;
    private Handler handler;

    @Before
    public void setUp() {
        logger = mock(Logger.class);
        handler = mock(Handler.class);

        PowerMockito.mockStatic(LogManager.class);
        LogManager logManager = mock(LogManager.class);
        when(LogManager.getLogManager()).thenReturn(logManager);
        when(logManager.getLogger(any())).thenReturn(logger);
    }

    @Test
    public void shouldAddAndroidHandlerIfSetup() {
        Log.setup();
        verify(logger).addHandler(any(AndroidHandler.class));
    }

    @Test
    public void shouldAddHandler() {
        Log.addHandler(handler);
        verify(logger).addHandler(handler);
    }

    @Test
    public void shouldRemoveHandler() {
        Log.removeHandler(handler);
        verify(logger).removeHandler(handler);
    }
}

File: library/src/test/kotlin/io/github/dmitrikudrenko/logger2/LogSetupTest.kt
package io.github.dmitrikudrenko.logger2

import io.github.dmitrikudrenko.logger2.impl.AndroidHandler
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Matchers
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import java.util.logging.Formatter
import java.util.logging.Handler
import java.util.logging.Logger

@RunWith(RobolectricTestRunner::class)
@Config(manifest = Config.NONE)
class LogSetupTest {
    private var logger: Logger? = null
    private var handler: Handler? = null
    private var androidHandler: AndroidHandler? = null
    private var logcatFormatter: Formatter? = null
    private var formatter: Formatter? = null

    @Before
    fun `set up`() {
        logger = mock(Logger::class.java)
        handler = mock(Handler::class.java)
        androidHandler = mock(AndroidHandler::class.java)
        logcatFormatter = mock(Formatter::class.java)
        formatter = mock(Formatter::class.java)

        Log.setDefaultHandler(androidHandler)
        Log.setFormatter(logcatFormatter, formatter)
        Log.setLogger(logger)
    }

    @After
    fun `tear down`() {
        Log.setDefaultHandler(null)
        Log.setFormatter(null, null)
        Log.setLogger(null)
    }

    @Test
    fun `should add android handler if setup`() {
        Log.setup()
        verify<Logger>(logger).addHandler(Matchers.any(AndroidHandler::class.java))
    }

    @Test
    fun `should add handler`() {
        Log.addHandler(handler)
        verify<Logger>(logger).addHandler(handler)
    }

    @Test
    fun `should remove handler`() {
        Log.removeHandler(handler)
        verify<Logger>(logger).removeHandler(handler)
    }

    @Test
    fun `should set formatter to new handler`() {
        Log.addHandler(handler)
        verify<Handler>(handler).formatter = Matchers.any(Formatter::class.java)
    }

    @Test
    fun `should set logcat formatter if setup`() {
        Log.setup()
        verify<AndroidHandler>(androidHandler).formatter = logcatFormatter
    }

    @Test
    fun `should set formatter if add handler`() {
        Log.addHandler(handler)
        verify<Handler>(handler).formatter = formatter
    }
}
